compiler/luci/service/src/CircleShapeInferenceRule.cpp

```diff
@@ -1023,6 +1023,27 @@ class ShapeInferenceAlgorithm final : public luci::CircleNodeVisitor<loco::NodeS
     return loco::NodeShape{input_shape};
   }
 
+  loco::NodeShape visit(const luci::CircleMatrixSetDiag *node) final
+  {
+    auto input_shape = loco::shape_get(node->input()).as<loco::TensorShape>();
+    auto diagonal_shape = loco::shape_get(node->diagonal()).as<loco::TensorShape>();
+
+    auto rank = diagonal_shape.rank();
+
+    LUCI_ASSERT(rank == input_shape.rank() - 1, "diagonal rank = input rank - 1");
+
+    for (uint32_t i = 0; i < rank - 1; i++)
+    {
+      LUCI_ASSERT(diagonal_shape.dim(i) == input_shape.dim(i), "diagonal dims = input dims");
+    }
+
+    auto dim = std::min(input_shape.dim(rank - 1).value(), input_shape.dim(rank).value());
+
+    LUCI_ASSERT(dim == diagonal_shape.dim(rank - 1), "Max diag len error");
```

Comment:
I am not sure about our conventions, maybe @seanshpark can tell more, but for me this looks like an overkill =)

I propose to move checks in separate function that returns `true` if everything is ok.
Leave one line, like `LUCI_ASSERT(check_function(input_shape, diagonal_shape), "Wrong diagonal shape")` in this function.

If we do so, this function will have only 3 lines, which is better for readability. (and could help avoid "Unused variable" compilation errors in case we redefine `LUCI_ASSERT`)

Reply:
I think it's both is OK, add assert for validations and remove assert for readability.
Just for information, some strange real models do have some violations and these asserts help to find them :)